[clang] report inlining decisions with -Wattribute-{warning|error}

nickdesaulniers opened the pull request:
Due to inlining, descovering which specific call site to a function with
the attribute "warning" or "error" is painful.

In the IR record inlining decisions in a new metadata tuple "inlined.from" when inlining a callee
that itself contains a call to a dontcall-error or dontcall-warn fn.

Print this info so that it's clearer which call site is problematic.

There's still some limitations with this approach; macro expansion is
not recorded.

Fixes: https://github.com/ClangBuiltLinux/linux/issues/1571

Differential Revision: https://reviews.llvm.org/D141451

nickdesaulniers committed:
[clang] report inlining decisions with -Wattribute-{warning|error}

Due to inlining, descovering which specific call site to a function with
the attribute "warning" or "error" is painful.

In the IR record inlining decisions in metadata when inlining a callee
that itself contains a call to a dontcall-error or dontcall-warn fn.

Print this info so that it's clearer which call site is problematic.

There's still some limitations with this approach; macro expansion is
not recorded.

Fixes: https://github.com/ClangBuiltLinux/linux/issues/1571

Differential Revision: https://reviews.llvm.org/D141451

nickdesaulniers commented:
(Initial import from https://reviews.llvm.org/D141451)

nickdesaulniers committed:
use structured bindings

addresses https://reviews.llvm.org/D141451#inline-1445382

nickdesaulniers committed:
use LocCookie

addresses https://reviews.llvm.org/D141451#inline-1445409

nickdesaulniers committed:
fix indentation of test

addresses https://reviews.llvm.org/D141451#inline-1445383

nickdesaulniers committed:
use < rather than -o -

Addresses https://reviews.llvm.org/D141451#inline-1445410

nickdesaulniers committed:
fix label checks

addresses https://reviews.llvm.org/D141451#inline-1445385

nickdesaulniers committed:
remove dead lines from test

addresses https://reviews.llvm.org/D141451#inline-1445386

nickdesaulniers committed:
return SmallVector

hopefully NRVO kicks in *shrug*

addresses https://reviews.llvm.org/D141451#inline-1446132

nickdesaulniers committed:
no need to support lone MDStrings; they're always enclosed in an MDTuple

addresses https://reviews.llvm.org/D141451#inline-1446137

nickdesaulniers committed:
use an Intanium mangled identifier in test

addresses https://reviews.llvm.org/D141451#inline-1445411

erichkeane commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -93,6 +93,8 @@ def err_fe_backend_error_attr :
 def warn_fe_backend_warning_attr :
   Warning<"call to '%0' declared with 'warning' attribute: %1">, BackendInfo,
   InGroup<BackendWarningAttributes>;
+def note_fe_backend_in : Note<"called by function '%0'">;
```
Is this the right file for this?  It seems weird to do codegen in a 'frontend kinds' diagnostics list, though I'm not convinced I know enough to ask this question well enough.

nickdesaulniers replied:
Where should the diagnostic strings go if not here?

erichkeane replied:
IDK, that is why I asked.  It SOUNDS wrong, so I'm hopeful someone else can come along and tell us.

nickdesaulniers replied:
All of the diagnostics for backend warnings are defined in this file. (see existing lines 66-95).

As is, this note is emitted alongside the err/warn from existing lines 91-95, so it makes sense to put the new notes immediately after those existing definitions.

If my newly added backend warning notes belong in a new file, then all of the above belong in a new file.  But that seems orthogonal to my patch.

nickdesaulniers commented:
Note to reviewers, I imported this to a GH PR from https://reviews.llvm.org/D141451, and addressed comments from https://reviews.llvm.org/D141451#4311399 down.

I did have a related RFC, but the [proposal](https://discourse.llvm.org/t/rfc-improving-clangs-middle-and-back-end-diagnostics/69261/16?u=nickdesaulniers) from @rnk is a yak shave that I don't intend to pursue.

Upon rereviewing https://reviews.llvm.org/D141451, it seems there was just some minor nits to resolve.  Bringing this back up because @kees reminded me how painful this is.

erichkeane commented on clang/lib/CodeGen/CodeGenAction.cpp:
```diff
@@ -794,6 +796,13 @@ void BackendConsumer::DontCallDiagHandler(const DiagnosticInfoDontCall &D) {
                               : diag::warn_fe_backend_warning_attr)
       << llvm::demangle(D.getFunctionName()) << D.getNote();
+
+  SmallVector<std::string> InliningDecisions = D.getInliningDecisions();
```
Could we use `StringRef` here instead?  It seems to make more sense to me.

nickdesaulniers replied:
done in e7ed56d009a83f78ffe22eaae8256a766b0c6b4b PTAL

erichkeane commented on clang/lib/CodeGen/CodeGenAction.cpp:
```diff
@@ -794,6 +796,13 @@ void BackendConsumer::DontCallDiagHandler(const DiagnosticInfoDontCall &D) {
+
+  SmallVector<std::string> InliningDecisions = D.getInliningDecisions();
+  InliningDecisions.push_back(D.getCaller().str());
+  for (auto [index, value] : llvm::enumerate(InliningDecisions))
```
```suggestion
  for (const auto &[index, value] : llvm::enumerate(InliningDecisions))
```

Seems we don't modify either, and `value` is a string, so requiring a copy of the pair is expensive.

nickdesaulniers replied:
done in d3a221a9aea4f60244adc7d726b398abeb2775b3 PTAL

erichkeane reviewed (commented):
only a few comments on the CFE, most of the work seems LLVM related.

nickdesaulniers committed:
use const ref structured binding

addresses https://github.com/llvm/llvm-project/pull/73552/files#r1406619873

nickdesaulniers committed:
s/std::string/StringRef/ for inlining decisions

addresses https://github.com/llvm/llvm-project/pull/73552/files#r1406619278

nickdesaulniers committed:
remove unnecessary include

nickdesaulniers committed:
Revert "no need to support lone MDStrings; they're always enclosed in an MDTuple"

This reverts commit a1496ded6a6aa43bc8d0329982156ef2f62321d2.

nickdesaulniers committed:
Revert "use LocCookie"

This reverts commit 239ab3ede24f030c40eb1e6cafc3dac6955835b6.

kees commented on llvm/lib/IR/DiagnosticInfo.cpp:
```diff
@@ -455,18 +455,11 @@ void DiagnosticInfoDontCall::print(DiagnosticPrinter &DP) const {
-    } else if (auto *S = dyn_cast<MDString>(MO)) {
-      InliningDecisions.push_back(S->getString().str());
-    }
-  }
```
We want it to read like "before", yes?

nickdesaulniers replied:
yes; note:
1. github is having some kind of outtage currently. https://www.githubstatus.com/ currently mentions an issue with PRs.  I have pushed a revert but it's not yet being reflected in the GH UI.
2. I need to fix the issue pointed out https://reviews.llvm.org/D141451#inline-1446137, which I haven't done yet. The revert was incorrect.

dwblaikie commented:
> I did have a related RFC, but the [proposal](https://discourse.llvm.org/t/rfc-improving-clangs-middle-and-back-end-diagnostics/69261/16?u=nickdesaulniers) from @rnk is a yak shave that I don't intend to pursue.

I'd push on that a bit more. As I mentioned on the phab review, I think it'd be unfortunate to build a second system for tracking inlining apart from the debug info.

Guess it comes down to whether some sufficient quorum insist on that being necessary yak shaving here or not.

nickdesaulniers commented on llvm/lib/IR/DiagnosticInfo.cpp:
```diff
@@ -449,3 +451,22 @@ void DiagnosticInfoDontCall::print(DiagnosticPrinter &DP) const {
+
+  if (MDN) {
+    const MDOperand &MO = MDN->getOperand(0);
+    if (auto *MDT = dyn_cast<MDTuple>(MO)) {
```
Copying @aeubanks 's [comment](https://reviews.llvm.org/D141451#inline-1446137) from the phab review:

> seems simpler if this is always a MDTuple instead of special casing one entry to be MDString

For the life of me I cannot figure out why when an MDTuple is created via:
```c++
Metadata *MD = MDString::get(CI->getContext(), CalledFunc->getName());
MDTuple *MDT = MDNode::get(CI->getContext(), {MD});
CI->setMetadata("inlined.from", MDT);
```
you get
```llvm
!10 = !{!"my_memcpy"}
```

but when you copy a `MDTuple` and then append to it, you get a layer of indirection:
```c++
Metadata *MD = MDString::get(CI->getContext(), CalledFunc->getName());                                                                                                                                                   
if (MDNode *N = CI->getMetadata("inlined.from")) {                                                                                                                                                                       
  TempMDTuple Temp = cast<MDTuple>(N)->clone();                                                                                                                                                                          
  Temp->push_back(MD);                                                                                                                                                                                                   
  MD = MDNode::replaceWithUniqued(std::move(Temp));                                                                                                                                                                      
}                                                                                                                                                                                                                        
MDTuple *MDT = MDNode::get(CI->getContext(), {MD});                                                                                                                                                                      
CI->setMetadata("inlined.from", MDT);
```
```llvm
!10 = !{!11}
!11 = !{!"my_memcpy", !"my_driver"}
```
when I would have expected:
```llvm
!10 = !{!"my_memcpy", !"my_driver"}
```
am I holding the API wrong??
